tests/test_eia.py

```diff
@@ -0,0 +1,645 @@
+        # basic test
+        today = datetime.today().replace(tzinfo=pytz.utc)
+        for ba in self.us_bas:
+            with self.assertRaises(ValueError):
+                self._run_net_test(ba, start_at=today+timedelta(days=1),
+                                   end_at=today+timedelta(days=2))
+
+    def test_get_trade_naive_start_at(self):
+        for ba in self.us_bas:
+            if ba in self.problem_bas_trade:
+                continue
+            self._run_net_test(ba, start_at="20161212", end_at="20161222T04Z")
+
+    def test_get_trade_naive_end_at(self):
+        for ba in self.us_bas:
+            if ba in self.problem_bas_trade:
+                continue
+            self._run_net_test(ba, start_at="20161212T04Z", end_at="20161222")
+
+    def test_get_trade_yesterday(self):
+        for ba in self.no_delay_bas:
+            if ba in self.problem_bas_trade:
+                continue
+            self._run_net_test(ba, yesterday=True)
+
+    def test_all_us_bas(self):
+        for ba in self.us_bas:
+            data = self._run_net_test(ba, market=self.MARKET_CHOICES.hourly)
+            self.assertGreater(len(data), 1)
+
+    def test_non_us_bas_raise_valueerror(self):
+        for ba in self.can_mex:
+            with self.assertRaises(ValueError):
+                self._run_net_test(ba, market=self.MARKET_CHOICES.hourly)
+
+    def _run_test(self, ba_name, expect_data=True, tol_min=0, **kwargs):
+        # set up
+        c = client_factory(ba_name)
+        # get data
+        data = c.get_trade(**kwargs)
+
+        # test number
+        if expect_data:
+            self.assertGreaterEqual(len(data), 1)
+        else:
+            self.assertEqual(data, [])
+
+        # test contents
+        for dp in data:
+            # test key names
+            self.assertEqual(set(['load_MW', 'ba_name',
+                                  'timestamp', 'freq', 'market']),
+                             set(dp.keys()))
+
+            # test values
+            self.assertEqual(dp['timestamp'].tzinfo, pytz.utc)
+            self.assertIn(dp['ba_name'], self.BA_CHOICES)
+
+            # test for numeric gen
+            self.assertGreaterEqual(dp['load_MW']+1, dp['load_MW'])
+
+            # test correct temporal relationship to now
+            if c.options['forecast']:
+                self.assertGreaterEqual(dp['timestamp'],
+                                        pytz.utc.localize(datetime.utcnow())-timedelta(minutes=tol_min))
+            else:
+                self.assertLess(dp['timestamp'], pytz.utc.localize(datetime.utcnow()))
+
+            # test within date range
+            start_at = c.options.get('start_at', False)
+            end_at = c.options.get('end_at', False)
+            if start_at and end_at:
+                self.assertGreaterEqual(dp['timestamp'], start_at)
+                self.assertLessEqual(dp['timestamp'], end_at)
+
+        # return
+        return data
+
+    # this is repeated- super it?
```

Comment:
Done for _run_test and _run_null_response_test. I'll see if there are others I can combine to clean things up a bit.